### PR TITLE
Add download option to 'Console output', move 'View as plain text' and 'Copy' buttons to app bar

### PR DESCRIPTION
**Pull request by janfaracik**
**Before**
image

**After**
image

Small MR to add a 'Download' link to the console output, as well as to move the controls to the app bar so that they're grouped. Let me know your thoughts on the updated placement - happy to discuss alternatives.

### Testing done

* Copy button works as before
* View as plain text works as before
* Download button downloads the full log

### Proposed changelog entries

- Add download option to 'Console output', move 'View as plain text' and 'Copy' buttons to app bar

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@jenkinsci/sig-ux 

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/resources/hudson/model/Run/console.jelly
Comment (mawinter69) [inline core/src/main/resources/hudson/model/Run/console.jelly:42]: Is it really useful to use the displayname here? If I choose a displayname that contains a `/` this wouldn't be saveable.
something
`${it.parent.name}-{it.number}.txt` would guarantee a valid filename

**Review by mawinter69 — CHANGES_REQUESTED** [review]
Is the copy option really the best thing as being the one that is directly accessible?
So far I never had really the need to copy the logs. But looking at the log in plain text I do from time to time

**Comment by mawinter69** [comment]
> Is the copy option really the best thing as being the one that is directly accessible? So far I never had really the need to copy the logs. But looking at the log in plain text I do from time to time

probably OK to have it like this.

File: core/src/main/resources/hudson/model/Run/console.jelly
Comment (daniel-beck) [reply core/src/main/resources/hudson/model/Run/console.jelly:42]: > If I choose a displayname that contains a `/` this wouldn't be saveable.

Per https://html.spec.whatwg.org/multipage/links.html#attr-hyperlink-download

> The attribute may have a value; the value, if any, specifies the default filename that the author recommends for use in labeling the resource in a local file system. There are no restrictions on allowed values, but authors are cautioned that most file systems have limitations with regard to what punctuation is supported in filenames, and user agents are likely to adjust filenames accordingly.

Trying this PR and DOM editing in Firefox, I get the file `foo_bar_#1.txt` from the attribute value `foo/bar/#1.txt`.

(`parent.fullName` might still be better though. Unsure whether we should go for display names, there's arguments for and against IMO)

**Comment by uhafner** [comment]
It took me a while to understand that the result is visible in freestyle jobs only. Isn't the PR that unifies the console in pipelines and freestyle jobs already merged?

I think it is a good idea to move the actions to the app bar. I'm not sure why the copy action is a button (I never use this action, does it copy the whole log or the visible section?), but I very often use the "raw console" link. Maybe we can show the latter as button as well (or only this one). The app bar has a lot of space that is not used yet. Do we have anonymous user statistics about such preferences?

**Comment by mawinter69** [comment]
> but I very often use the "raw console" link

same for me

**Comment by daniel-beck** [comment]
> I'm not sure why the copy action is a button (I never use this action, does it copy the whole log or the visible section?)

My guess is because it's a recently added action (#8960) that has been previously implemented like this? Copying seems like the least needed use case, and between view raw, and download raw, even seems redundant. (And it only copies the visible section, via JS.)

**Comment by janfaracik** [comment]
> > I'm not sure why the copy action is a button (I never use this action, does it copy the whole log or the visible section?)
> 
> My guess is because it's a recently added action (#8960) that has been previously implemented like this? Copying seems like the least needed use case, and between view raw, and download raw, even seems redundant. (And it only copies the visible section, via JS.)

Essentially that - I didn't want to undo the existing work, happy to prioritise the order of download/complete log however as I agree.

**Comment by janfaracik** [comment]
ATH passing - https://github.com/jenkinsci/acceptance-test-harness/pull/1563

**Review by timja — COMMENTED** [review]
Needs a matching PR to Pipeline (probably workflow-job) as well.

Apart from that LGTM

File: core/src/main/resources/hudson/model/Run/console.jelly
Comment (mawinter69) [inline core/src/main/resources/hudson/model/Run/console.jelly:31]: the dropdown lib is no longer needed

**Review by mawinter69 — APPROVED** [review]
looks good. One minor thing but not blocking

**Comment by janfaracik** [comment]
> Needs a matching PR to Pipeline (probably workflow-job) as well.
> 
> Apart from that LGTM

Done - thanks! https://github.com/jenkinsci/workflow-job-plugin/pull/439

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by FredericSouchon** [comment]
Hello,

as I told in JENKINS-73657, This is a real regression because it is requires to open the "console output" view that may crash the browser if the log is heavy enough. this use case may have been forgotten.

Could you submit the comeback of the "view as plain text" option in the lateral menu ?

Thanks in advance,

**Comment by JanKirsch** [comment]
Hi there, same here.

> This is a real regression because it is requires to open the "console output" view that may crash the browser if the log is heavy enough. this use case may have been forgotten.

Yes, that doesn't seem to have been taken into account. On our Jenkins we have huge logs, which leads to a broken "Console Output" view at approx. 80% of our jobs. We have to wait to get to the now hidden "View as plain text" button, because it is not clickable until the browser is able to breath again. That makes me very sad. I totally understand why this change come up, but unfortunately the "Console Output" view is not stable enough to handle huge logs.

> Could you submit the comeback of the "view as plain text" option in the lateral menu ?

Please bring it back or come up with a configurable Job "Navigation menu" or improve the "Console Output" view. No matter what you do, please make the community happy again :)

**Comment by mawinter69** [comment]
Could be easily done with a plugin.

**Comment by mawinter69** [comment]
https://github.com/jenkins-infra/repository-permissions-updater/issues/4063

**Comment by mirao** [comment]
Since 2.462.1, we can see the bug https://issues.jenkins.io/browse/JENKINS-73683.
As this pull request contains some changes related to console output, is it possible that it caused the issue with timestamps as mentioned in that bug?

**Comment by MarkEWaite** [comment]
> Since [2.462.1](https://www.jenkins.io/changelog-stable/#v2.462.1), we can see the bug https://issues.jenkins.io/browse/JENKINS-73683. As this pull request contains some changes related to console output, is it possible that it caused the issue with timestamps as mentioned in that bug?

It seems unlikely to me, but you can experiment with it by building a Jenkins core yourself with this pull request reverted.  See the [CONTRIBUTING file](https://github.com/jenkinsci/jenkins/blob/master/CONTRIBUTING.md#building-and-debugging) for the instructions to build Jenkins core yourself.

**Comment by shubhamm05** [comment]
With this new feature, as soon we open the build. only option is to first open console logs in browser. 
when logs size is huge the browser gets stuck or tab kills.
The option to view logs in plain text comes only after opening the logs. 
earlier, we had option to open directly as plain text without opening the logs.
please see if view in plain text can be moved back like earlier.

**Comment by mawinter69** [comment]
@shubhamm05 
https://plugins.jenkins.io/plain-text-console-link/ adds the link back to the side panel

**Pull request merged**
